pkg/action/executor/horizontal_scaler.go

```diff
@@ -0,0 +1,308 @@
+		Content(content).
+		Create()
+	glog.V(4).Infof("Horizontal scaling action is built as %v", action)
+
+	return &action, nil
+}
+
+func (h *HorizontalScaler) getProviderPod(actionItem *proto.ActionItemDTO) (*api.Pod, error) {
+	targetEntityType := actionItem.GetTargetSE().GetEntityType()
+	var providerPod *api.Pod
+	if targetEntityType == proto.EntityDTO_CONTAINER_POD {
+		targetPod := actionItem.GetTargetSE()
+		id := targetPod.GetId()
+		foundPod, err := util.GetPodFromIdentifier(h.kubeClient, id)
+		if err != nil {
+			return nil, err
+		}
+		providerPod = foundPod
+	} else if targetEntityType == proto.EntityDTO_APPLICATION {
+		providers := actionItem.GetProviders()
+
+		foundPod, err := util.FindApplicationPodProvider(h.kubeClient, providers)
+		if err != nil {
+			return nil, err
+		}
+		providerPod = foundPod
+	} else if targetEntityType == proto.EntityDTO_VIRTUAL_APPLICATION {
+		// TODO, unbind action is send as MOVE
+		// An example actionItemDTO is as follows:
+
+		// actionItemDTO:<actionType:MOVE uuid:"_YaxeQDPxEea2efRvkwVM-Q"
+		// targetSE:<entityType:VIRTUAL_APPLICATION id:"vApp-apache2" displayName:"vApp-apache2" > currentSE:<
+		// entityType:APPLICATION id:"apache2::default/frontend-hhp6g" displayName:"apache2::default/frontend-
+		// hhp6g" > providers:<entityType:APPLICATION ids:"apache2::default/frontend-57ytg" ids:"apache2::
+		// default/frontend-ue0kg" ids:"apache2::default/frontend-hhp6g" ids:"apache2::default/frontend-luohe">>
+		//
+		// TODO find the pod name based on application ID. App id is in the following format.
+		// TODO, maybe use service related info to find the corresponding pod.
+		// ProcessName::PodNamespace/PodName
+		currentSE := actionItem.GetCurrentSE()
+		targetEntityType := currentSE.GetEntityType()
+		if targetEntityType != proto.EntityDTO_APPLICATION {
+
+		}
+		appName := currentSE.GetId()
+		ids := strings.Split(appName, "::")
+		if len(ids) < 2 {
+			return nil, fmt.Errorf("%s is not a valid Application ID. Unbind failed.", appName)
+		}
+		podIdentifier := ids[1]
+
+		// Get the target pod from podIdentifier.
+		pod, err := util.GetPodFromIdentifier(h.kubeClient, podIdentifier)
+		if err != nil {
+			return nil, err
+		}
+		providerPod = pod
+	}
+	return providerPod, nil
+}
+
+func (h *HorizontalScaler) horizontalScale(action *turboaction.TurboAction) (*turboaction.TurboAction, error) {
+	// 1. Setup consumer
+	actionContent := action.Content
+	scaleSpec, ok := actionContent.ActionSpec.(turboaction.ScaleSpec)
+	if !ok || scaleSpec.NewReplicas < 0 {
+		return nil, errors.New("Failed to setup horizontal scaler as the provided scale spec is invalid.")
+	}
+
+	var key string
+	if actionContent.ParentObjectRef.ParentObjectUID != "" {
+		key = actionContent.ParentObjectRef.ParentObjectUID
+	} else {
+		return nil, errors.New("Failed to setup horizontal scaler consumer: failed to retrieve the key.")
```

Comment:
Change it to failed to retrieve the UUID of RC or Deployment, since key is the var name which is a bit confusing